mllib/src/main/scala/org/apache/spark/ml/feature/TargetEncoder.scala

```diff
@@ -0,0 +1,409 @@
+
+  /**
+   * Param for how to handle invalid data during transform().
+   * Options are 'keep' (invalid data presented as an extra categorical feature) or
+   * 'error' (throw an error).
+   * Note that this Param is only used during transform; during fitting, invalid data
+   * will result in an error.
+   * Default: "error"
+   * @group param
+   */
+  @Since("4.0.0")
+  override val handleInvalid: Param[String] = new Param[String](this, "handleInvalid",
+    "How to handle invalid data during transform(). " +
+    "Options are 'keep' (invalid data presented as an extra categorical feature) " +
+    "or error (throw an error). Note that this Param is only used during transform; " +
+    "during fitting, invalid data will result in an error.",
+    ParamValidators.inArray(TargetEncoder.supportedHandleInvalids))
+
+  setDefault(handleInvalid -> TargetEncoder.ERROR_INVALID)
+
+  @Since("4.0.0")
+  val targetType: Param[String] = new Param[String](this, "targetType",
+    "How to handle invalid data during transform(). " +
+      "Options are 'keep' (invalid data presented as an extra categorical feature) " +
+      "or error (throw an error). Note that this Param is only used during transform; " +
+      "during fitting, invalid data will result in an error.",
+    ParamValidators.inArray(TargetEncoder.supportedTargetTypes))
+
+  setDefault(targetType -> TargetEncoder.TARGET_BINARY)
+
+  final def getTargetType: String = $(targetType)
+
+  @Since("4.0.0")
+  val smoothing: DoubleParam = new DoubleParam(this, "smoothing",
+    "lower bound of the output feature range",
+    ParamValidators.gtEq(0.0))
+
+  setDefault(smoothing -> 0.0)
+
+  final def getSmoothing: Double = $(smoothing)
+
+  private[feature] lazy val inputFeatures = if (isSet(inputCol)) Array($(inputCol))
+                                            else if (isSet(inputCols)) $(inputCols)
+                                            else Array.empty[String]
+
+  private[feature] lazy val outputFeatures = if (isSet(outputCol)) Array($(outputCol))
+                                    else if (isSet(outputCols)) $(outputCols)
+                                    else inputFeatures.map{field: String => s"${field}_indexed"}
+
+  private[feature] def validateSchema(schema: StructType,
+                                      fitting: Boolean): StructType = {
+
+    require(inputFeatures.length > 0,
+      s"At least one input column must be specified.")
+
+    require(inputFeatures.length == outputFeatures.length,
+      s"The number of input columns ${inputFeatures.length} must be the same as the number of " +
+        s"output columns ${outputFeatures.length}.")
+
+    val features = if (fitting) inputFeatures :+ $(labelCol)
+                  else inputFeatures
+
+    features.foreach {
+      feature => {
+        try {
+          val field = schema(feature)
+          if (field.dataType != DoubleType) {
```

Comment:
i mimic this behavior from other encoders (i.e. OneHotEncoder)
what would be your approach?  accepting Integers?  checking for nominal attribute in metadata?

Reply:
I think that's OK if it's what other encoders do. But I see checks like https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/feature/OneHotEncoder.scala#L93 - maybe follow that?

Reply:
you're right
now accepting any subclass of NumericType for features & label
(maybe it doesn't make much sense the continuous case, it could be done anyway)